File: lib_network/src/main/java/com/android/sdk/net/ErrorMessageFactoryImpl.java
package com.android.sdk.net;

import android.text.TextUtils;

import androidx.annotation.NonNull;

import com.android.sdk.net.core.exception.ApiErrorException;
import com.android.sdk.net.core.exception.ErrorMessageFactory;
import com.android.sdk.net.core.exception.ServerErrorException;
import com.android.sdk.net.core.provider.ErrorMessageConverter;
import com.android.sdk.net.rxjava2.RxJavaChecker;

import java.io.IOException;
import java.util.NoSuchElementException;

import retrofit2.HttpException;
import timber.log.Timber;

/**
 * @author Ztiany
 */
final class ErrorMessageFactoryImpl implements ErrorMessageFactory {

    @NonNull
    @Override
    public CharSequence createMessage(@NonNull Throwable exception) {
        ErrorMessageConverter errorMessageConverter = NetContext.get().commonConfig().errorMessageConverter();

        Timber.d("createMessage with：%s", exception.toString());

        CharSequence message = null;
        //1：Network Connection Error Processing
        if (exception instanceof IOException) {
            message = errorMessageConverter.netErrorMessage(exception);
        }

        //2：Server Error Processing
        else if (exception instanceof ServerErrorException) {
            int errorType = ((ServerErrorException) exception).getErrorType();
            if (errorType == ServerErrorException.DATA_PARSE_ERROR) {
                message = errorMessageConverter.serverDataParseErrorMessage(exception);
            } else if (errorType == ServerErrorException.EMPTY_SERVER_DATA) {
                message = errorMessageConverter.nullEntityErrorMessage(exception);
            }
        }

        //3：Http Response Code is not 200
        else if (exception instanceof HttpException) {
            int code = ((HttpException) exception).code();
            if (code >= 500/*http 500*/) {
                message = errorMessageConverter.serverInternalErrorMessage(exception);
            } else if (code >= 400/*http 400*/) {
                message = errorMessageConverter.clientRequestErrorMessage(exception);
            }
        }

        //4：Api Error
        else if (exception instanceof ApiErrorException) {
            message = exception.getMessage();
            if (TextUtils.isEmpty(message)) {
                message = errorMessageConverter.apiErrorMessage((ApiErrorException) exception);
            }
        }

        //5：RxJava Single
        else if (RxJavaChecker.hasRxJava2() && exception instanceof NoSuchElementException) {
            message = errorMessageConverter.serverInternalErrorMessage(exception);
        }

        //6：Others
        if (isEmpty(message)) {
            message = errorMessageConverter.unknownErrorMessage(exception);
        }

        return message;
    }

    private static boolean isEmpty(CharSequence str) {
        return str == null || str.toString().trim().isEmpty();
    }

}

File: lib_network/src/main/java/com/android/sdk/net/impl/ErrorMessageFactoryImpl.kt
package com.android.sdk.net.impl

import android.text.TextUtils
import com.android.sdk.net.NetContext
import com.android.sdk.net.core.exception.ApiErrorException
import com.android.sdk.net.core.config.ErrorMessageFactory
import com.android.sdk.net.core.exception.ServerErrorException
import retrofit2.HttpException
import java.io.IOException

/**
 * @author Ztiany
 */
internal class ErrorMessageFactoryImpl : ErrorMessageFactory {

    override fun convert(exception: Throwable): CharSequence {
        val errorMessageConverter = NetContext.get().commonConfig().errorMessageConverter()

        var message: CharSequence? = null
        if (exception is IOException) {
            message = errorMessageConverter.convertWhenNetError(exception)
        } else if (exception is ServerErrorException) {
            val errorType = exception.errorType
            if (errorType == ServerErrorException.DATA_PARSE_ERROR) {
                message = errorMessageConverter.convertWhenParsingDataFailed(exception)
            } else if (errorType == ServerErrorException.EMPTY_SERVER_DATA) {
                message = errorMessageConverter.convertWhenNoDataReturned(exception)
            }
        } else if (exception is HttpException) {
            val code = exception.code()
            if (code >= 500 /*http 500*/) {
                message = errorMessageConverter.convertWhenServerInternalError(exception)
            } else if (code >= 400 /*http 400*/) {
                message = errorMessageConverter.convertWhenClientRequestFailed(exception)
            }
        } else if (exception is ApiErrorException) {
            message = exception.message
            if (TextUtils.isEmpty(message)) {
                message = errorMessageConverter.convertWhenApiException(exception)
            }
        }

        return message.takeIf {
            !it.isNullOrEmpty()
        } ?: errorMessageConverter.convertWhenUnknownError(exception)
    }

}